### PR TITLE
Double check validations on learners tested

### PR DESCRIPTION
**Pull request by gsvr**
https://praekelt.atlassian.net/browse/RZ-20?jql=project%20%3D%20RZ%20AND%20resolution%20%3D%20Unresolved%20AND%20priority%20%3D%20Critical%20ORDER%20BY%20key%20DESC

### CODE REVIEW

**Comment by gsvr** [comment]
@imsickofmaps Please review this, I'll add girls' validation if you're happy with the boys'.  This way if not approved I don't have to rework.

File: js_sandbox/lib/go-rts-zambia.js [hunk 1]
```diff
@@ -330,6 +330,42 @@ function GoRtsZambia() {
...
+        var value_arr = [];
+        for (var i = 0; i < state_name_arr.length; i++) {
+            value_arr[i] = parseInt(im.get_user_answer(state_name_arr[i]),10);
+        }
+        value_arr.push(parseInt(content,10));
+        return value_arr;
+    };
+
+    self.craft_error_msg = function(array_total, gender, array_of_answers, total) {
+        var l1 = "You've entered results for ";
```
Comment (imsickofmaps) [inline js_sandbox/lib/go-rts-zambia.js]: Use `['first', 'second'].join(' ')` instead

File: js_sandbox/lib/go-rts-zambia.js [hunk 2]
```diff
@@ -1403,203 +1452,289 @@ function GoRtsZambia() {
...
+                // think it will create a problem the way i've hacked it
+                self.add_state(self.make_totals_error_state("error_state_1", "perf_learner_boys_total", msg));
+                return "error_state_1";
+            }
+        },
+        "In total, how many boys achieved 16 out of 20 or more?",
         function(content) {
             // check that the value provided is actually decimal-ish.
-            return self.check_valid_number(content);
+            return self.check_valid_number(content) && (parseInt(content,10) >= 0);
```
Comment (imsickofmaps) [inline js_sandbox/lib/go-rts-zambia.js]: Change all to `self.check_valid_number(content)`

File: js_sandbox/lib/go-rts-zambia.js [hunk 3]
```diff
@@ -584,7 +621,16 @@ function GoRtsZambia() {
         );
     };
 
-    // END Shared creators
+    self.make_totals_error_state = function(state_name, next_state, error_msg) {
```
Comment (gsvr) [inline js_sandbox/lib/go-rts-zambia.js:629]: @hodgestar This is my helper for creating a new ChoiceState

File: js_sandbox/lib/go-rts-zambia.js [hunk 4]
```diff
@@ -1403,203 +1452,287 @@ function GoRtsZambia() {
...
+            var array_of_answers = self.make_array_of_answers(content, boys_states_completed);
+            var array_total = self.calc_array_total(array_of_answers);
+
+            if(array_total < boys_total) {
+                return "perf_learner_boys_desirable_results";
+            } else {
+                var msg = self.craft_error_msg(array_total, "boys", array_of_answers, boys_total);
+                // askmike: i'm having problems with re-using the state creator with the same state_name,
+                // think it will create a problem the way i've hacked it
+                self.add_state(self.make_totals_error_state("error_state_1", "perf_learner_boys_total", msg));
```
Comment (gsvr) [inline js_sandbox/lib/go-rts-zambia.js]: @hodgestar This is an example usage of my helper.  The tests pass now, but if it tries to run twice it bombs (trying to add_state that already exists)

Comment (hodgestar) [inline js_sandbox/lib/go-rts-zambia.js]: I think `make_totals_error_state` is fine, but the way this is being used isn't. Turn this into an `add_creator` and return the error state if content creation fails.

**Comment by gsvr** [comment]
@imsickofmaps Ready for review

File: js_sandbox/lib/go-rts-zambia.js [hunk 5]
```diff
@@ -584,7 +626,72 @@ function GoRtsZambia() {
...
+        } else {
+            total_state = "perf_learner_girls_total";
+        }
+
+        var total = parseInt(im.get_user_answer(total_state),10);
+        var array_of_answers = self.make_array_of_answers(states_completed);
+        var array_total = self.calc_array_total(array_of_answers);
+        var msg;
+
+        if (must_equal) {
```
Comment (imsickofmaps) [inline js_sandbox/lib/go-rts-zambia.js]: You can DRY this with:
`if (must_equal && (total === array_total) || !must_equal && (total > array_total)) {
    # make freetext
} else {
    # make error state
}`

see: https://gist.github.com/imsickofmaps/049af3fdd23d05787bb6

Comment (gsvr) [inline js_sandbox/lib/go-rts-zambia.js]: I just did - not sure if you're finished with review but pushed the change.

**Comment by imsickofmaps** [comment]
Looks good, only missing a test to check that Boys error loops back to boys start again, and same for girls.

**Comment by gsvr** [comment]
Ok, should be good to go now.

**Comment by imsickofmaps** [comment]
:+1:

**Comment by gsvr** [comment]
@imsickofmaps Should we still merge and close this?

**Comment by imsickofmaps** [comment]
@gsvr yup - I left it so you can do `git flow feature finish` locally to avoid the "working on the same branch still" issue from before.

**Comment by gsvr** [comment]
:+1: done :)

**Comment by imsickofmaps** [comment]
You still need to push the develop branch so it closes this issue because it sees  the commits merged

**Pull request merged**